### PR TITLE
D8MT-224 - oe_migration_views

### PR DESCRIPTION
**Pull request by vever001**
### Description

Adds the OpenEuropa Migration Views module allows you to create Views from `migrate_map_*` and `migrate_message_*` tables.
See [README.md](https://github.com/openeuropa/oe_migration/blob/D8MT-224/modules/oe_migration_views/README.md)

### CODE REVIEW

File: modules/oe_migration_views/oe_migration_views.views.inc [hunk 1]
```diff
@@ -0,0 +1,468 @@
...
+  $migrations = $plugin_manager->createInstances([]);
+
+  /** @var \Drupal\migrate\Plugin\Migration $migration */
+  foreach ($migrations as $migration) {
+    /** @var \Drupal\migrate\Plugin\migrate\id_map\Sql $id_map */
+    $id_map = $migration->getIdMap();
+    if (!is_a($id_map, Sql::class)) {
+      continue;
+    }
+
```
Comment (samuelsolis) [inline modules/oe_migration_views/oe_migration_views.views.inc:41]: This function cause a problem installing the module when others migrate modules are installed (and is more problematic when a lot of migration are defined, you can try with migrate_drupal, for example).

And I think that this is the problem that cause the error installing the site with the docker compose environment. This function increase the size of the $data array, that is saved in cache (and I was investigating but we cannot avoid it) and cause the error in the CacheSet method. I don't have the final solution but I think that this lines can helps:

```
   // Note only for the PR review:  Comment from the README.
    // Views relationships are available to join entity tables if the migrate destination plugin is an `EntityContentBase`.
    if (!is_subclass_of($migration->getDestinationPlugin(), 'Drupal\migrate\Plugin\migrate\destination\Entity')) {
      continue;
    }
```

Basically, as I read in the README this lines remove from the array the content unrelated with EntityContentBase. The problem is that these lines improve the code and reduce the size of $data but it's still failing, we still have tons of lines of $data ready to be saved.

Check the code in ViewsData.php:259, where the view is cached. 

I'm not sure about the requirement of this module and the purpose of expose this huge amount of data into a view, but here is the problem. It's too much and it's a problem saving data in cache due the size of the serialized array.

Comment (vever001) [reply modules/oe_migration_views/oe_migration_views.views.inc:41]: Hi @samuelsolis 
This function is basically exposing the migration tables to views. As there can be a lot of migrations from various modules the size of that array can become quite big.
As for the filtering on migration which use EntityContentBase as destination, it's already [there](https://github.com/openeuropa/oe_migration/blob/D8MT-224/modules/oe_migration_views/oe_migration_views.views.inc#L292).

I made a small fix yesterday in the PR which should reduce the size of the array.
Could you give it a try on your project and see if that makes a difference pls?
If it's not enough I guess I'll have to do something with config to expose only certain migrations to views.

Comment (samuelsolis) [reply modules/oe_migration_views/oe_migration_views.views.inc:41]: Hi again @vever001 

I tried to install the site again (after the new pushed changes) but the module still throwing the same error. Debugging I have notice that the array still have 522 elements and I think that it's difficult to reduce it without some config. I think that we should create a kind of "entities exposed to the migration views" or something like that to filter the number of elements... It's difficult to manage it with tons of data, it's an exponencial growing and difficult to reduce without a "manual" filter.

Comment (vever001) [reply modules/oe_migration_views/oe_migration_views.views.inc:41]: Yes, thanks for checking. I was kinda expecting this, with migrate_drupal and other migration modules the amount of migrations can become quite big.
By default it seems max_allowed_packet is set to 4MB. So unless we increase it which of course is not ideal, config to expose only specific migrations to views can't be avoided I think (though even then, 4MB might be too low, depending on the amount of existing views and migrations).

Just so I understand correctly, do you still use this module for OE/DMT migrations? Is the views integration still required or do you only export the migration messages (to e.g. CSV) and check the reports there?
As it can become a bit tedious to maintain migrations and migration report views...
I just want to make sure the requirement haven't changed since the last time I worked on this.
Let me know

Comment (samuelsolis) [reply modules/oe_migration_views/oe_migration_views.views.inc:41]: I really don't know the purpose of this module and why we are developing this or if we need it. I asked Ioannis about the requirement of this module (in order to perform a better peer review, and to check if the code is valid for the requirements) but we don't have requirements nor specifications, I don't know why you are the person who has to develop it and no the DMT team (I suppose that we have a reason, for sure, but I don't know it) and really, I don't know nothing. 

I only know that if we enable the module, as you say, with migrate_drupal and something more, the array is huge and the memory limit is not enough, and we can't increase the memory limit (as far as I know). 

Maybe we need to re-define the task before continue.

File: modules/oe_migration_views/src/Plugin/views/field/MigrateMessagesField.php [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+    }
+
+    $source_ids_hash = $this->getValue($values);
+    $result = $this->getMigrateMessages($this->configuration['message_table'], $source_ids_hash);
+    $levels = _oe_migration_views_migrate_message_level_options();
+
+    $rows = [];
+    foreach ($result as $message_row) {
+      $level = isset($levels[$message_row->level]) ? $levels[$message_row->level] : $this->t('Unknown');
+      $rows[] = strtoupper($level) . ': ' . $message_row->message;
```
Comment (ioannis-ampatzis) [inline modules/oe_migration_views/src/Plugin/views/field/MigrateMessagesField.php:35]: `$level` is an object and when it's passed as parameter to `strtoupper()` it throws a `TypeError` exception. This only happens when there are messages to display.
Probably what you meant to do was `strtoupper($level->getUntranslatedString())`

File: modules/oe_migration_views/oe_migration_views.views.inc [hunk 3]
```diff
@@ -0,0 +1,480 @@
...
+ * @throws \Drupal\migrate\MigrateException
+ * @throws \Drupal\migrate\MigrateSkipRowException
+ */
+function _oe_migration_views_views_data_alter_migrate_map_relationship_definitions(array &$data, Migration $migration) {
+  $id_map = $migration->getIdMap();
+  $map_table = $id_map->mapTableName();
+
+  $destination_plugin = $migration->getDestinationPlugin();
+  if (is_a($destination_plugin, EntityContentBase::class)) {
+    [, $entity_type_id] = explode(':', $destination_plugin->getPluginId());
```
Comment (ioannis-ampatzis) [inline modules/oe_migration_views/oe_migration_views.views.inc:293]: `[, $entity_type_id]` causes a "parser has reached an invalid state" warning.

File: modules/oe_migration_views/oe_migration_views.views.inc [hunk 4]
```diff
@@ -0,0 +1,480 @@
...
+use Drupal\migrate\Plugin\migrate\destination\EntityContentBase;
+use Drupal\migrate\Plugin\migrate\id_map\Sql;
+use Drupal\migrate\Plugin\Migration;
+
+/**
+ * Implements hook_views_data_alter().
+ */
+function oe_migration_views_views_data_alter(&$data) {
+  $plugin_manager = \Drupal::service('plugin.manager.migration');
+  $migrations = $plugin_manager->createInstances([]);
```
Comment (ioannis-ampatzis) [inline modules/oe_migration_views/oe_migration_views.views.inc:22]: If the source database is not installed/accessible, as it actually happens with the containers in acceptance and prod, it will through two exceptions and will stop the execution of the pipeline.
```
In Connection.php line 420:
  SQLSTATE[HY000] [1049] Unknown database '[d7 database]'  

In Connection.php line 416:
  SQLSTATE[HY000] [1049] Unknown database '[d7 database]' 
```
